File: integration_tests/java_client_tests/client-app/src/test/java/dev/graphlink/test/CacheTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.client.JacksonGraphLinkJsonCodec;
import dev.graphlink.test.generated.enums.UserStatus;
import dev.graphlink.test.generated.inputs.AddressInput;
import dev.graphlink.test.generated.inputs.CreateUserInput;
import dev.graphlink.test.generated.types.GetCachedUserResponse;
import dev.graphlink.test.generated.types.GetStaleUserResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class CacheTest {

    private MockAdapter adapter;
    private GraphLinkClient client;

    private static final AddressInput kAddressInput = AddressInput.builder()
            .street("123 Main St").city("Springfield").country("US").zip("12345").build();

    private static final CreateUserInput kCreateInput = CreateUserInput.builder()
            .name("Alice Smith").email("alice@test.com")
            .status(UserStatus.ACTIVE).address(kAddressInput).build();

    @BeforeEach
    void setUp() {
        adapter = new MockAdapter();
        client = newClient(adapter, new MockWebSocketAdapter());
    }

    // ── Cache hit ─────────────────────────────────────────────────────────────

    @Nested
    class CacheHit {

        @BeforeEach
        void register() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
        }

        @Test
        void firstCall_hitsAdapter() {
            client.queries.getCachedUser("user-1");
            assertEquals(1, adapter.getCallCount());
        }

        @Test
        void secondCallSameArg_servedFromCache_noExtraAdapterCall() {
            client.queries.getCachedUser("user-1");
            client.queries.getCachedUser("user-1");
            assertEquals(1, adapter.getCallCount());
        }

        @Test
        void cachedResponse_hasCorrectData() {
            client.queries.getCachedUser("user-1");
            GetCachedUserResponse res = client.queries.getCachedUser("user-1");
            assertNotNull(res.getGetCachedUser());
            assertEquals("user-1", res.getGetCachedUser().getId());
        }

        @Test
        void differentArg_bypassesCache() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserBobJson));
            client.queries.getCachedUser("user-1");
            client.queries.getCachedUser("user-2");
            assertEquals(2, adapter.getCallCount());
        }
    }

    // ── Tag-based invalidation ────────────────────────────────────────────────

    @Nested
    class TagInvalidation {

        @Test
        void createCachedUser_invalidatesUsersTag_nextGetCachedUserRefetches() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            adapter.registerData("createCachedUser", Map.of("createCachedUser", kUserAliceJson));

            client.queries.getCachedUser("user-1");       // populates cache
            assertEquals(1, adapter.getCallCount());

            client.mutations.createCachedUser(kCreateInput); // invalidates "users"
            assertEquals(2, adapter.getCallCount());

            client.queries.getCachedUser("user-1");       // cache was cleared — must re-fetch
            assertEquals(3, adapter.getCallCount());
        }

        @Test
        void resetAll_invalidatesAllTags_allCachedQueriesRefetch() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            adapter.registerData("getCachedPost", Map.of("getCachedPost", kPostJson));
            adapter.registerData("resetAll", Map.of("resetAll", true));

            client.queries.getCachedUser("user-1");
            client.queries.getCachedPost("post-1");
            assertEquals(2, adapter.getCallCount());

            client.mutations.resetAll();                  // invalidates all
            assertEquals(3, adapter.getCallCount());

            client.queries.getCachedUser("user-1");       // re-fetches
            client.queries.getCachedPost("post-1");       // re-fetches
            assertEquals(5, adapter.getCallCount());
        }
    }

    // ── Cross-tag isolation ───────────────────────────────────────────────────

    @Nested
    class CrossTagIsolation {

        @Test
        void usersTagInvalidation_doesNotAffect_postsCache() {
            adapter.registerData("getCachedPost", Map.of("getCachedPost", kPostJson));
            adapter.registerData("createCachedUser", Map.of("createCachedUser", kUserAliceJson));

            client.queries.getCachedPost("post-1");      // cache "posts"
            assertEquals(1, adapter.getCallCount());

            client.mutations.createCachedUser(kCreateInput); // invalidates "users" only
            assertEquals(2, adapter.getCallCount());

            client.queries.getCachedPost("post-1");      // "posts" cache intact — no extra call
            assertEquals(2, adapter.getCallCount());
        }

        @Test
        void transferPost_invalidatesBothTags() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            adapter.registerData("getCachedPost", Map.of("getCachedPost", kPostJson));
            adapter.registerData("transferPost", Map.of("transferPost", kPostJson));

            client.queries.getCachedUser("user-1");
            client.queries.getCachedPost("post-1");
            assertEquals(2, adapter.getCallCount());

            client.mutations.transferPost("post-1", "user-2"); // invalidates "users" + "posts"
            assertEquals(3, adapter.getCallCount());

            client.queries.getCachedUser("user-1");      // must re-fetch
            client.queries.getCachedPost("post-1");      // must re-fetch
            assertEquals(5, adapter.getCallCount());
        }
    }

    // ── listCachedUsers — same tag group as getCachedUser ─────────────────────

    @Nested
    class ListCachedUsers {

        @Test
        void listCachedUsers_isCachedAfterFirstCall() {
            adapter.registerData("listCachedUsers",
                    Map.of("listCachedUsers", List.of(kUserAliceJson)));
            client.queries.listCachedUsers();
            client.queries.listCachedUsers();
            assertEquals(1, adapter.getCallCount());
        }

        @Test
        void createCachedUser_invalidatesListCachedUsers() {
            adapter.registerData("listCachedUsers",
                    Map.of("listCachedUsers", List.of(kUserAliceJson)));
            adapter.registerData("createCachedUser", Map.of("createCachedUser", kUserAliceJson));

            client.queries.listCachedUsers();             // populate cache
            client.mutations.createCachedUser(kCreateInput); // invalidate
            client.queries.listCachedUsers();             // re-fetch
            assertEquals(3, adapter.getCallCount());
        }
    }

    // ── staleIfOffline ────────────────────────────────────────────────────────

    @Nested
    class StaleIfOffline {

        @Test
        void staleData_isReturnedWhenNetworkFails_afterTtlExpires() throws InterruptedException {
            adapter.registerData("getStaleUser", map("getStaleUser", kUserAliceJson));

            client.queries.getStaleUser("user-1");       // populate cache (TTL = 1s)
            assertEquals(1, adapter.getCallCount());

            Thread.sleep(1100);                           // wait for TTL to expire

            adapter.simulateFailure = true;
            GetStaleUserResponse res = client.queries.getStaleUser("user-1");

            assertNotNull(res.getGetStaleUser());
            assertEquals("user-1", res.getGetStaleUser().getId());
        }

        @Test
        void nonStaleEntry_throwsWhenNetworkFails() {
            adapter.registerData("getCachedUser", Map.of("getCachedUser", kUserAliceJson));
            // Don't warm the cache — fresh request fails immediately
            adapter.simulateFailure = true;
            assertThrows(RuntimeException.class,
                    () -> client.queries.getCachedUser("user-1"));
        }
    }
}

File: integration_tests/kotlin_client_tests/client-app/src/test/kotlin/dev/graphlink/test/CacheTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.enums.UserStatus
import dev.graphlink.test.generated.inputs.AddressInput
import dev.graphlink.test.generated.inputs.CreateUserInput
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.kPostJson
import dev.graphlink.test.Fixtures.kUserAliceJson
import dev.graphlink.test.Fixtures.kUserBobJson
import dev.graphlink.test.Fixtures.newClient
import dev.graphlink.test.n

class CacheTest {

    private lateinit var adapter: MockAdapter
    private lateinit var client: GraphLinkClient

    private val kAddressInput = AddressInput("123 Main St", "Springfield", "US", "12345")
    private val kCreateInput = CreateUserInput(
        name = "Alice Smith",
        email = "alice@test.com",
        status = UserStatus.ACTIVE,
        address = kAddressInput,
    )

    @BeforeTest
    fun setUp() {
        adapter = MockAdapter()
        client = newClient(adapter, MockWebSocketAdapter())
    }

    // ── Cache hit ─────────────────────────────────────────────────────────────

    @Nested
    inner class CacheHit {

        @BeforeTest
        fun register() {
            adapter.registerData("getCachedUser", mapOf("getCachedUser" to kUserAliceJson))
        }

        @Test
        fun firstCall_hitsAdapter() = runTest {
            client.queries.getCachedUser("user-1")
            assertEquals(1, adapter.getCallCount())
        }

        @Test
        fun secondCallSameArg_servedFromCache_noExtraAdapterCall() = runTest {
            client.queries.getCachedUser("user-1")
            client.queries.getCachedUser("user-1")
            assertEquals(1, adapter.getCallCount())
        }

        @Test
        fun cachedResponse_hasCorrectData() = runTest {
            client.queries.getCachedUser("user-1")
            val res = client.queries.getCachedUser("user-1")
            assertNotNull(res.getCachedUser)
            assertEquals("user-1", res.getCachedUser.id)
        }

        @Test
        fun differentArg_bypassesCache() = runTest {
            adapter.registerData("getCachedUser", mapOf("getCachedUser" to kUserBobJson))
            client.queries.getCachedUser("user-1")
            client.queries.getCachedUser("user-2")
            assertEquals(2, adapter.getCallCount())
        }
    }

    // ── Tag-based invalidation ────────────────────────────────────────────────

    @Nested
    inner class TagInvalidation {

        @Test
        fun createCachedUser_invalidatesUsersTag_nextGetCachedUserRefetches() = runTest {
            adapter.registerData("getCachedUser", mapOf("getCachedUser" to kUserAliceJson))
            adapter.registerData("createCachedUser", mapOf("createCachedUser" to kUserAliceJson))

            client.queries.getCachedUser("user-1")
            assertEquals(1, adapter.getCallCount())

            client.mutations.createCachedUser(kCreateInput)
            assertEquals(2, adapter.getCallCount())

            client.queries.getCachedUser("user-1")
            assertEquals(3, adapter.getCallCount())
        }

        @Test
        fun resetAll_invalidatesAllTags_allCachedQueriesRefetch() = runTest {
            adapter.registerData("getCachedUser", mapOf("getCachedUser" to kUserAliceJson))
            adapter.registerData("getCachedPost", mapOf("getCachedPost" to kPostJson))
            adapter.registerData("resetAll", mapOf("resetAll" to true))

            client.queries.getCachedUser("user-1")
            client.queries.getCachedPost("post-1")
            assertEquals(2, adapter.getCallCount())

            client.mutations.resetAll()
            assertEquals(3, adapter.getCallCount())

            client.queries.getCachedUser("user-1")
            client.queries.getCachedPost("post-1")
            assertEquals(5, adapter.getCallCount())
        }
    }

    // ── Cross-tag isolation ───────────────────────────────────────────────────

    @Nested
    inner class CrossTagIsolation {

        @Test
        fun usersTagInvalidation_doesNotAffect_postsCache() = runTest {
            adapter.registerData("getCachedPost", mapOf("getCachedPost" to kPostJson))
            adapter.registerData("createCachedUser", mapOf("createCachedUser" to kUserAliceJson))

            client.queries.getCachedPost("post-1")
            assertEquals(1, adapter.getCallCount())

            client.mutations.createCachedUser(kCreateInput)
            assertEquals(2, adapter.getCallCount())

            client.queries.getCachedPost("post-1")
            assertEquals(2, adapter.getCallCount())
        }

        @Test
        fun transferPost_invalidatesBothTags() = runTest {
            adapter.registerData("getCachedUser", mapOf("getCachedUser" to kUserAliceJson))
            adapter.registerData("getCachedPost", mapOf("getCachedPost" to kPostJson))
            adapter.registerData("transferPost", mapOf("transferPost" to kPostJson))

            client.queries.getCachedUser("user-1")
            client.queries.getCachedPost("post-1")
            assertEquals(2, adapter.getCallCount())

            client.mutations.transferPost("post-1", "user-2")
            assertEquals(3, adapter.getCallCount())

            client.queries.getCachedUser("user-1")
            client.queries.getCachedPost("post-1")
            assertEquals(5, adapter.getCallCount())
        }
    }

    // ── listCachedUsers ───────────────────────────────────────────────────────

    @Nested
    inner class ListCachedUsers {

        @Test
        fun listCachedUsers_isCachedAfterFirstCall() = runTest {
            adapter.registerData("listCachedUsers", mapOf("listCachedUsers" to listOf(kUserAliceJson)))
            client.queries.listCachedUsers()
            client.queries.listCachedUsers()
            assertEquals(1, adapter.getCallCount())
        }

        @Test
        fun createCachedUser_invalidatesListCachedUsers() = runTest {
            adapter.registerData("listCachedUsers", mapOf("listCachedUsers" to listOf(kUserAliceJson)))
            adapter.registerData("createCachedUser", mapOf("createCachedUser" to kUserAliceJson))

            client.queries.listCachedUsers()
            client.mutations.createCachedUser(kCreateInput)
            client.queries.listCachedUsers()
            assertEquals(3, adapter.getCallCount())
        }
    }

    // ── staleIfOffline ────────────────────────────────────────────────────────

    @Nested
    inner class StaleIfOffline {

        @Test
        fun staleData_isReturnedWhenNetworkFails_afterTtlExpires() = runTest {
            adapter.registerData("getStaleUser", n("getStaleUser" to kUserAliceJson))

            client.queries.getStaleUser("user-1")
            assertEquals(1, adapter.getCallCount())

            @Suppress("BlockingMethodInNonBlockingContext")
            Thread.sleep(1100) // wait for TTL to expire (TTL = 1s, uses real time)

            adapter.simulateFailure = true
            val res = client.queries.getStaleUser("user-1")

            assertNotNull(res.getStaleUser)
            assertEquals("user-1", res.getStaleUser.id)
        }

        @Test
        fun nonStaleEntry_throwsWhenNetworkFails() = runTest {
            adapter.registerData("getCachedUser", mapOf("getCachedUser" to kUserAliceJson))
            adapter.simulateFailure = true
            assertFailsWith<Exception> {
                client.queries.getCachedUser("user-1")
            }
        }
    }
}
